inc/classes/class-rocket-critical-css.php

```diff
@@ -266,4 +311,308 @@ public function set_items() {
+			return $buffer;
+		}
+	
+		$noscripts = '';
+	
+		foreach ( $tags_match[0] as $i => $tag ) {
+			// Strip query args.
+			$path = rocket_extract_url_component( $tags_match[2][ $i ], PHP_URL_PATH );
+	
+			// Check if this file should be deferred.
+			if ( isset( $excluded_css[ $path ] ) ) {
+				continue;
+			}
+	
+			$preload = str_replace( 'stylesheet', 'preload', $tags_match[1][ $i ] );
+			$onload  = str_replace( $tags_match[3][ $i ], ' as="style" onload=""' . $tags_match[3][ $i ] . '>', $tags_match[3][ $i ] );
+			$tag     = str_replace( $tags_match[3][ $i ] . '>', $onload, $tag );
+			$tag     = str_replace( $tags_match[1][ $i ], $preload, $tag );
+			$tag     = str_replace( 'onload=""', 'onload="this.rel=\'stylesheet\'"', $tag );
+			$buffer  = str_replace( $tags_match[0][ $i ], $tag, $buffer );
+	
+			$noscripts .= '<noscript>' . $tags_match[0][ $i ] . '</noscript>';
+		}
+	
+		$buffer = str_replace( '</body>', $noscripts . '</body>', $buffer );
+	
+		return $buffer;
+	}
+
+	/**
+	 * Insert critical CSS in the <head>
+	 *
+	 * @since 2.10
+	 * @author Remy Perona
+	 */
+	function insert_critical_css() {
+		global $pagenow;
+	
+		if ( ! get_rocket_option( 'async_css' ) ) {
+			return;
+		}
+	
+		if ( is_rocket_post_excluded_option( 'async_css' ) ) {
+			return;
+		}
+	
+		// Don't apply on wp-login.php/wp-register.php.
+		if ( in_array( $pagenow, array( 'wp-login.php', 'wp-register.php' ), true ) ) {
+			return;
+		}
+	
+		if ( ( defined( 'DONOTROCKETOPTIMIZE' ) && DONOTROCKETOPTIMIZE ) || ( defined( 'DONOTASYNCCSS' ) && DONOTASYNCCSS ) ) {
+			return;
+		}
+	
+		// Don't apply if user is logged-in and cache for logged-in user is off.
+		if ( is_user_logged_in() && ! get_rocket_option( 'cache_logged_user' ) ) {
+			return;
+		}
+	
+		// This filter is documented in inc/front/process.php.
+		$rocket_cache_search = apply_filters( 'rocket_cache_search', false );
+	
+		// Don't apply on search page.
+		if ( is_search() && ! $rocket_cache_search ) {
+			return;
+		}
+	
+		// Don't apply on excluded pages.
+		if ( in_array( $_SERVER['REQUEST_URI'] , get_rocket_option( 'cache_reject_uri' , array() ), true ) ) {
+			return;
+		}
+	
+		// Don't apply on 404 page.
+		if ( is_404() ) {
+			return;
+		}
+	
+		if ( is_home() ) {
+			$name = 'home.css';
+		} elseif ( is_front_page() ) {
+			$name = 'front_page.css';
+		} elseif ( is_category() ) {
+			$name = 'category.css';
+		} elseif ( is_tag() ) {
+			$name = 'post_tag.css';
+		} elseif ( is_tax() ) {
+			$taxonomy = get_queried_object()->term_name;
+			$name = $taxonomy . '.css';
+		} elseif ( is_singular() ) {
+			$post_type = get_post_type();
+			$name = $post_type . '.css';
+		} else {
+			$name = 'front_page.css';
+		}
+
+		$file = $this->critical_css_path . $name;
+	
+		if ( ! rocket_direct_filesystem()->is_readable( $file ) ) {
+			return;
+		}
+	
+		echo '<style id="rocket-critical-css">' . rocket_direct_filesystem()->get_contents( $file ) . '</style>';
```

Comment:
@Screenfeed I feel this should be sanitized even though it's read from a file, which function would be best?

Reply:
>You can use wp_kses( $data, array( "\'", '\"' ) ); or wp_strip_all_tags( $data ); or esc_html( $data ); but this is NOT 100% secure. The actual (sep. 2015) best way to sanitize custom CSS rules is to use CSSTidy.

Source: [Function Reference/wp_add_inline_style](https://codex.wordpress.org/Function_Reference/wp_add_inline_style).
(Now I know where does `wp_kses( $data, array( "\'", '\"' ) )` come from.)

No perfect solution it seems. Personally I'd use `wp_kses()` when saving, and `wp_strip_all_tags( esc_html( $data ) )` when displaying.

Reply:
Edit: I meant `esc_html( wp_strip_all_tags( $data ) )`.

Reply:
thank you, changes done